src/shmem_internal.h

```diff
@@ -285,6 +285,32 @@ extern unsigned int shmem_internal_rand_seed;
         }                                                                               \
     } while (0)
 
+/* Indicates whether the memory region pointed to by ptr1 (having size1) overlaps with
+ * the buffer at ptr2 (having size2).
+ *   SHMEM usage examples:
+ *     For the collective routines (src/dest arguments):
+ *         size1 == size2 and complete_overlap_allowed is 1
+ *     For the pt2pt sync routines (ivars/status/indices arguments):
+ *         size1 != size2 and complete_overlap_allowed is 0 */
+#define SHMEM_ERR_CHECK_OVERLAP(ptr1, ptr2, size1, size2, complete_overlap_allowed)     \
+    do {                                                                                \
+        const void *p1 = (void*)(ptr1);                                                 \
+        const void *p2 = (void*)(ptr2);                                                 \
+        const void *ptr_low  = p1 > p2 ? p2 : p1;                                       \
+        const void *ptr_high = p1 > p2 ? p1 : p2;                                       \
+        const size_t sz_low  = p1 > p2 ? size2 : size1;                                 \
+        const void *ptr_extent = (void *)((char *)ptr_low + sz_low);                    \
+        if (complete_overlap_allowed && p1 == p2) {                                     \
+            break; /* Skip this check when buffer is allowed to completely overlap  */  \
+        }                                                                               \
+        if (ptr_extent > ptr_high) {                                                    \
```

Comment:
Is `ptr_extent > ptr_high + sz_high` a complete overlap case?

Reply:
Good question - my understanding of the OpenSHMEM spec is that it's only a complete overlap when the two buffers start at the same address and "have the same size" (but there is only 1 size argument in APIs where complete overlap is allowed: the RMA and collective routines).

I'm not sure the spec says that explicitly though - is your interpretation different @wrrobin?  Do I understand your concern correctly?

Reply:
Yes, thanks @wrrobin.  This is resolved and I corrected a small mistake in my response above: complete overlap is not allowed in RMA.